### PR TITLE
apply flagos te_groups_gemm op

### PR DESCRIPTION
**Pull request by chai-xiaonan**
- add ```te_general_grouped_gemm``` op for flagos backend, base on flag_gems
- support both forward and backward computation, distinguished by ```grad```

### CODE REVIEW

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 1]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+        D = []
+        for i in range(num_gemms):
+            m = A[i].shape[1] if transa else A[i].shape[0]
+            n = B[i].shape[0] if transb else B[i].shape[1]
+            D.append(torch.empty((m, n), dtype=D[i].dtype, device=A[0].device))
+
+    def gelu_backward(grad_output, x):
+        # Approximation of GELU derivative commonly used in Transformer Engine
+        cdf = 0.5 * (1.0 + flag_gems.erf(x / flag_gems.sqrt(2.0)))
+        pdf = flag_gems.exp(-0.5 * x * x) / flag_gems.sqrt(2.0 * math.pi)
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: import math

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 2]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+    for i in range(num_gemms):
+        # Handle the special case of zero-element inputs
+        if A[i].numel() == 0 or B[i].numel() == 0:
+            if not single_output:
+                if D[i].numel() != 0 and not accumulate:
+                    D[i].zero_()
+            else:
+                out = torch.zeros(A[i].shape[0], B[i].shape[1])
+            if grad and len(bias) > i and bias[i] is not None and bias[i].numel() != 0:
+                bias[i].zero_()
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: bias or bias_grad?

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 3]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+
+    num_gemms = len(A)
+    if D is None:
+        D = []
+        for i in range(num_gemms):
+            m = A[i].shape[1] if transa else A[i].shape[0]
+            n = B[i].shape[0] if transb else B[i].shape[1]
+            D.append(torch.empty((m, n), dtype=D[i].dtype, device=A[0].device))
+
+    def gelu_backward(grad_output, x):
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: why not use flag_gems.gelu_backward

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 4]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
         return D, bias_grad, gelu_input, extra_output_ret
     else:
         return out1, bias_grad, gelu_input, extra_output_ret
+
+
+def te_general_grouped_gemm_fl(
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py:124]: Please add a comment indicating that this function can represent both forward computation and backward computation, distinguished by the ```grad``` parameter.

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 1 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: ok

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 5]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+                bias_grad = out.sum(dim=0)
+                if accumulate:
+                    bias[i].add_(bias_grad)
+                else:
+                    bias[i].copy_(bias_grad)
+
+        if not single_output:
+            # Store output
+            if accumulate:
+                D[i].add_(out.to(D[i].dtype))
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: use ```flag_gems.to_copy``` to replace ```xxx.to```

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 2 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: Will add annotations

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 3 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: OK，Will attempt to replace and test.

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 4 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py:124]: OK

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 5 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: OK，Will attempt to replace and test.

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 6]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+    if single_output and D is None:
+        raise ValueError("not implemented, D should be allocated for single output case.")
+
+    num_gemms = len(A)
+    if D is None:
+        D = []
+        for i in range(num_gemms):
+            m = A[i].shape[1] if transa else A[i].shape[0]
+            n = B[i].shape[0] if transb else B[i].shape[1]
+            D.append(torch.empty((m, n), dtype=D[i].dtype, device=A[0].device))
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py:152]: why not use flag_gems.zeros

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 7]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+        pdf = flag_gems.exp(-0.5 * x * x) / flag_gems.sqrt(2.0 * math.pi)
+        return flag_gems.mul(grad_output, (cdf + x * pdf))
+
+    temp_D = []
+    for i in range(num_gemms):
+        # Handle the special case of zero-element inputs
+        if A[i].numel() == 0 or B[i].numel() == 0:
+            if not single_output:
+                if D[i].numel() != 0 and not accumulate:
+                    D[i].zero_()
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: try to use ```flag_gems.zeros```

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 8]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+            else:
+                out = torch.zeros(A[i].shape[0], B[i].shape[1])
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: try to use ```flag_gems.zeros```

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 9]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+            if grad and len(bias) > i and bias[i] is not None and bias[i].numel() != 0:
+                bias[i].zero_()
+            if (
+                len(pre_gelu_out) > i
+                and pre_gelu_out[i] is not None
+                and pre_gelu_out[i].numel() != 0
+            ):
+                pre_gelu_out[i].zero_()
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: try to use ```flag_gems.zeros```

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 10]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+        if not grad:
+            if has_bias:
+                # Fused matrix multiplication and bias addition
+                out = flag_gems.addmm(bias[i], a, b)
+            else:
+                out = flag_gems.mm(a, b)
+
+            # Apply GELU epilogue if pre_gelu_out is provided
+            if has_pre_gelu:
+                pre_gelu_out[i].copy_(out)
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: try to use ```flag_gems.copy```

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 11]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+                out = flag_gems.gelu(out)
+        else:
+            out = flag_gems.mm(a, b)
+            if has_pre_gelu:
+                out = gelu_backward(out, pre_gelu_out[i])
+
+            # Compute bias gradients if requested
+            if has_bias:
+                bias_grad = out.sum(dim=0)
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: try to use ```flag_gems.sum_dim```

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 12]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+                if accumulate:
+                    bias[i].add_(bias_grad)
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: try to use ```flag_gems.add```

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 13]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+                else:
+                    bias[i].copy_(bias_grad)
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: try to use ```flag_gems.copy```

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 14]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+
+        if not single_output:
+            # Store output
+            if accumulate:
+                D[i].add_(out.to(D[i].dtype))
+            else:
+                D[i].copy_(out.to(D[i].dtype))
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: as above

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 15]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+        else:
+            temp_D.append(out.to(D[0].dtype))
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py:215]: try to use ```flag_gems.to_copy```

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 6 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py:152]: ok

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 7 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: ok

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 8 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: ok

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 9 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: ok

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 10 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: ok

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 15 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py:215]: ok

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 14 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: ok

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 13 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: ok

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 12 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: ok

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py (hunk 11 above)
Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: ok

File: transformer_engine/plugin/core/backends/flagos/impl/gemm.py [hunk 16]
```diff
@@ -115,3 +116,107 @@ def generic_gemm_fl(
...
+        else:
+            out = flag_gems.mm(a, b)
+
+            # Apply dGELU epilogue if requested
+            if has_pre_gelu:
+                out = flag_gems.gelu_backward(out, pre_gelu_out[i])
+
+            # Compute bias gradients if requested
+            if has_bias:
+                bias_grad = flag_gems.sum_dim(out, dim=0)
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: please check ```dim``` argument of ```flag_gems.sum_dim```, is it should be a list or not

Comment (chai-xiaonan) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: In the `flag_gems.sum_dim` operator, the `dim `parameter supports multiple data types for input, including None, int, and List[int].

Comment (lxd-cumt) [reply transformer_engine/plugin/core/backends/flagos/impl/gemm.py]: please check
```
def sum_dim_comm(inp, dim=None, keepdim=False, *, dtype=None, out=None):
    if dtype is None:
        dtype = inp.dtype
        if dtype is torch.bool:
            dtype = torch.int64

    if dim == []:
        if not keepdim:
            return sum(inp, dtype=dtype)
        else:
            dim_num = inp.ndim
            return torch.reshape(sum(inp, dtype=dtype), [1] * dim_num)

    shape = list(inp.shape)
    dim = [d % inp.ndim for d in dim]
```

**Review by lxd-cumt — APPROVED** [review]
LGTM

File: transformer_engine/plugin/tests/test_te_general_grouped.py [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+
+                if grad or has_bias:
+                    for py_b, te_b in zip(bias_list_py, bias_list_te):
+                        self.assert_close(
+                            py_b, te_b, rtol=1e-3, atol=1e-3, msg="Bias gradient tensors mismatch!"
+                        )
+                print(f"    ✓ {backend_name}")
+            except NotImplementedError:
+                self.skipped += 1
+                print(f"    �~J~X {backend_name} (not implemented)")
```
Comment (lxd-cumt) [inline transformer_engine/plugin/tests/test_te_general_grouped.py]: please format

Comment (lxd-cumt) [inline transformer_engine/plugin/tests/test_te_general_grouped.py]: please format

**Review by zhaoyinglia — APPROVED** [review]
